lesson/outline/rtl-outline-template.md

```diff
@@ -0,0 +1,330 @@
+5. Once learners know how to query DOM nodes with RTL, go through a few different jest matchers provided by `@testing-library/jest-dom`. Explain that these matchers are extremely helpful when looking at the state of DOM nodes in our unit tests. E.g. whether or not a button is disabled, or a checkbox is checked, etc.
+
+    Once again, it is best to look these up on the [docs](https://github.com/testing-library/jest-dom) instead of memorizing them.
+```javascript
+expect(getByTestId('button')).toBeDisabled()
+expect(getByTestId('checkbox')).toBeChecked()
+```
+6. Demonstrate how RTL gives a readable error message when a DOM component is not present or when a test has failed. The debug message can be printed with `screen.debug()`.
+
+### Checkpoints Summary
+
+_Check out the [content standards](http://curriculum-documentation.codecademy.com/Content-Standards/checkpoint/) for guidance on writing narratives for exercises._
+
+1. Extract query methods from `render`
+2. Perform the different query types on component.
+- getByRole
+- queryByRole
+- findByRole
+3. Use the different query methods and jest matchers from `expect` to determine if a component is working as expected.
+
+#### What is the purpose of these checkpoints?
+1. So that learners understand how to perform queries on DOM nodes.
+2. So that scholars understand the difference between the different query types and know when to apply which.
+3. So that learners can apply the different jest matchers to verify the state of the DOM.
+
+### What would you like to have in the workspace for this exercise? Share your plan below.
+- A simple react component with some text.
+
+_Check out the [content standards](http://curriculum-documentation.codecademy.com/Content-Standards/workspaces/) for guidance on writing narratives for exercises._
+
+## Exercise # 4: Mimicking User interactions.
+
+### Which course outcomes will be covered by this exercise?
+
+1. Learners will be able mimick user interactions E.g. a mouseclick, keystrokes, etc.
+
+### Narrative Summary
+
+_Check out the [content standards](http://curriculum-documentation.codecademy.com/Content-Standards/narrative/) for guidance on writing narratives for exercises._
+
+1. RTL has different methods that can be used to mimick user interactions. E.g. `user-event` or `fireEvent`.
+```javascript
+test('type', () => {
+  render(<textarea />)
+
+  userEvent.type(screen.getByRole('textbox'), 'Hey There')
+  expect(screen.getByRole('textbox')).toHaveValue('Hey There')
+})
+```
+```javascript
+// Taken from RTL docs
+const Button = ({ onClick, children }) => (
+  <button onClick={onClick}>{children}</button>
+)
+
+test('calls onClick prop when clicked', () => {
+  const handleClick = jest.fn()
+  render(<Button onClick={handleClick}>Click Me</Button>)
+  fireEvent.click(screen.getByText(/click me/i))
```

Comment:
There's a preference for `userEvent` over `fireEvent`: https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-user-event.md